### ISSUE TITLE
Symlinks Reolve To Linux Root If Located In /Volumes/SystemRoot

### ISSUE THREAD

**Issue by TheBrokenRail** [darlinghq/darling#970]
**Expected Result**
Any symlink located in ``/Volumes/SystemRoot`` would behave just like any other symlink. A symlink targeting ``/usr/bin/python`` would resolve to Darling's Python.

**Actual Result**
If a symlink is located in ``/Volumes/SystemRoot`` and targets the root, it will resolve to Linux's root, not Darling's. For instance a symlink targeting ``/usr/bin/python`` located inside of ``/Volumes/SystemRoot`` will resolve to Linux's Python binary.

**Steps To Reproduce**
```
# Current Directory: /Users/user
Darling [~]$ ln -s /usr/bin/python pytest2
Darling [~]$ file pytest2
pytest2: Mach-O 64-bit executable x86_64 # Darling's Python
# Current Directory: /Volumes/SystemRoot/home/user/
Darling [/Volumes/SystemRoot/home/user]$ ln -s /usr/bin/python pytest2
Darling [/Volumes/SystemRoot/home/user]$ file pytest2
pytest2: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=df727276ffa30f0dd5d2ec239b1ea064370c98fe, for GNU/Linux 3.2.0, stripped # Linux's Python
```

**Additional Information**
This is an issue with the resolution of symlinks, not the creation. I know this because outside of Darling, both symlinks are identical:
```
user@user-ThinkPad-T495s:~$ file pytest2 
pytest2: symbolic link to /usr/bin/python
user@user-ThinkPad-T495s:~$ cd ~/.darling/Users/user/
user@user-ThinkPad-T495s:~/.darling/Users/user$ file pytest2 
pytest2: symbolic link to /usr/bin/python
```

XTrace Log: https://hatebin.com/hbanhkirky

**System Information**
Ubuntu 20.04

| Software | Version |
| --- | --- |
| Linux Kernel | ``5.8.0-50-generic`` |
| Darling | ``35534fcaa312b184145c95a543cf86c8b08d4e70`` |

**Comment by facekapow** [comment]
@LubosD Correct me if I'm wrong, but this is one of those "it's not a bug, it's a feature" things, right? i just tried to "fix" this locally and found that it causes `/lib64/ld-linux-x86-64.so.2` not to be found because it links absolutely to `/lib/x86_64-linux-gnu/ld-2.31.so`, which according to the "expected" behavior, should be resolved relative to the prefix (i.e. `/home/user/.darling/lib/x86_64-linux-gnu/ld-2.31.so`). That's definitely not what we in this case, however.

**Comment by bugaevc** [comment]
But that one is an ELF, it should not be resolved by the vprefix code at all. Can elfloader use raw open?

And it's wrong of your distro to use the absolute path in the symlink in this case.

**Comment by facekapow** [comment]
On second thought, you're right; this is definitely a bug (after all, I wouldn't expect Linux to read other drives' symlinks according to their roots). I agree that this does seem like the behavior we want; we certainly could make an exception for libelfloader and have it load the interpreter with non-vchrooted calls, and it doesn't seem like there's anywhere else we rely on this behavior.

And yes, it does seem awfully dumb of Ubuntu to provide an absolute path for the ld-linux symlink.

**Comment by facekapow** [comment]
I've pushed a potential fix onto the [`vchroot-abs-symlinks`](https://github.com/darlinghq/darling/tree/vchroot-abs-symlinks) branch (because I'm still not entirely sure if we want to have a special case for libelfloader; maybe there's another/better way).

**Comment by HinTak** [comment]
FWIW, there are plenty of absolute Symlinks on Linux, I believe.

**Comment by bugaevc** [comment]
Name some?

**Comment by HinTak** [comment]
You can use `find /usr  -type l -ls | grep -e '-> /'` to list some. I am piping it into `wc -l` and it is taking a long time.

But many under `/usr/sbin` (for security, I guess), many under mingw and wine.

finished - I have 9309 such under `/usr`, on fedora 34:

```
# find /usr  -type l -ls | grep -e '-> /' | wc -l
9309
```

**Comment by HinTak** [comment]
A lot under `/usr/bin` too, and `/usr/share/help`, etc. Anyway, run the above command (or variants of) to get a full list.

**Comment by HinTak** [comment]
The situation on a headless Raspbian (i.e. without X/GUI) is better but still 320 of these:

```
pi@raspberrypi:~ $ sudo find /usr  -type l -ls | grep -e '-> /' | wc -l
320
```

**Comment by HinTak** [comment]
On the headless Raspbian, most of the 320 are `/usr/sbin` and `/usr/bin`, and many under `/usr/share/man`, and disabled systemd services (`/usr/lib/systemd/system/ .... -> /dev/null` ), and quite a lot of libraries under `/usr/lib/arm-linux-gnueabihf/` (I think this is a Debian / Ubuntu convention, hence not seen on Fedora)

**Comment by bugaevc** [comment]
`-> /dev/null` ones are OK, but the rest... I'd say it's wrong of the distros to do that

**Comment by HinTak** [comment]
That's two mayor distros I referenced (Fedora and Raspbian / Debian), you are using a minor distro....

**Comment by bugaevc** [comment]
I _am_ using Fedora and Debian; doesn't mean that I agree with each of their decisions 🙂

**Comment by HinTak** [comment]
I have a Ubuntu server for Raspberrypi mounted too, and the situation with Ubuntu (headless) is similar to Raspbian / Debian:

```
find /run/media/Hin-Tak/writable/usr/  -type l -ls | greP -e '-> /' | wc -l
313
```

**Comment by HinTak** [comment]
The Ubuntu situation is very similar to Raspbian, except for `/usr/lib/aarch64-linux-gnu/` vs `/usr/lib/arm-linux-gnueabihf/` (since that's 64-bit Ubuntu vs 32-bit Raspbian).

**Comment by HinTak** [comment]
I think it definitely make sense for symlinks in `/usr/sbin` (and to a lesser extent /usr/bin) to be absolute for security reasons; mingw / wine probably do it to make sure their own chroot-like system works. The others I don't feel strongly either way. I am just saying it is quite common.

**Comment by bugaevc** [comment]
What security reasons?

**Comment by HinTak** [comment]
I am of course still concerned about darling recursing down the host file system - https://github.com/darlinghq/darling/issues/854 :(

**Comment by facekapow** [comment]
This has been fixed in 3761b3d7f1148e1a12cf41ea291504d2bb0fe6bb, which was recently merged into master.
